<a href="https://colab.research.google.com/github/ElBaron-27/UACJ_Ruben/blob/main/Actividad_Calificada_07_Datos_Temporales_271774.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos temporales**
### Limpieza y transformación de fechas, horas y duraciones en un dataset real

---

| | |
|---|---|
| **Materia** | Programación para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformación de datos: datos temporales |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Viajes de taxi en Nueva York, marzo de 2019 (NYC TLC, muestra publicada en un repositorio público de GitHub) |
| **Fecha de entrega** | [28.09.2026] |

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo** |Ruben Alejandro Uribe Rodriguez|
| **Matrícula** |271774 |

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu código en las celdas marcadas con `# Escribe tu código aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu código**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificación de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en tus resultados (cita cifras).
7. Usa los métodos vistos en la sesión: `pd.to_datetime()`, el accesor `.dt`, `pd.Timedelta`, `.between()`, `tz_localize()`, `groupby()`, `sort_values()`, `.where()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecución → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Temporales_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset contiene 6,433 viajes de taxis amarillos y verdes de la ciudad de Nueva York registrados en **marzo de 2019**. Cada fila es un viaje. Las fechas y horas están en **hora local de Nueva York**.

| Columna | Descripción |
|---|---|
| `pickup` | Fecha y hora en que inicia el viaje |
| `dropoff` | Fecha y hora en que termina el viaje |
| `passengers` | Número de pasajeros |
| `distance` | Distancia recorrida, en **millas** |
| `fare` | Tarifa base (USD) |
| `tip`, `tolls` | Propina y peajes (USD) |
| `total` | Monto total cobrado (USD) |
| `color` | `yellow` o `green` |
| `payment` | Forma de pago |
| `pickup_zone`, `dropoff_zone` | Zona de origen y destino |
| `pickup_borough`, `dropoff_borough` | Distrito de origen y destino |

> **Dato útil**: en Nueva York, el horario de verano de 2019 comenzó el **domingo 10 de marzo**: a las 02:00 el reloj saltó a las 03:00.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Conversión y validación** | 20 | 18–20: conversión con formato explícito y verificación de `NaT`; viajes fuera de periodo detectados; análisis del cambio de horario correcto; código comentado; conclusión con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o conclusión. | 1–13: conversión sin verificar o validaciones incompletas; conclusión sin sustento. | 0: no se entregó o el código no produce resultados. |
| **2. Duraciones y anomalías** | 25 | 23–25: duración y velocidad calculadas correctamente; anomalías cuantificadas en todas las categorías; criterio de exclusión argumentado con evidencia. | 18–22: cálculos correctos; criterio parcial o con poca evidencia. | 1–17: cálculos con errores o anomalías no cuantificadas. | 0: no se entregó o es incorrecto. |
| **3. Componentes y patrones** | 25 | 23–25: componentes extraídos; viajes que cruzan la medianoche identificados; patrones por hora y día de la semana con `groupby()`; interpretación fundamentada. | 18–22: componentes y agrupaciones correctos; interpretación incompleta. | 1–17: agrupaciones con errores o sin interpretación. | 0: no se entregó o es incorrecto. |
| **4. Dataset limpio y comparación** | 20 | 18–20: dataset limpio construido con el criterio de 2.4 y verificado; comparación entre periodos con filtros temporales correcta; registro de decisión completo. | 14–17: dataset limpio correcto; comparación o registro incompletos. | 1–13: criterio aplicado de forma inconsistente o sin verificación. | 0: no se entregó o es incorrecto. |
| **Reflexión final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relación con los resultados. | 0: no se entregó. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.3) y los comentarios del código cuentan para la calificación **de esa actividad**, no solo para la reflexión final.
- Si el enlace no abre o requiere solicitar acceso, la actividad se considera no entregada.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificación es 0. El descuento se aplica sobre la calificación obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [1]:
import pandas as pd

# Carga del dataset de taxis de Nueva York desde un repositorio público de GitHub.
# El archivo se lee con pandas; no se requiere ninguna otra biblioteca.
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/taxis.csv'
taxis = pd.read_csv(url)

print('Dimensiones:', taxis.shape)
taxis.head()

Dimensiones: (6433, 14)


,pickup,dropoff,passengers,distance,fare,tip,tolls,total,color,payment,pickup_zone,dropoff_zone,pickup_borough,dropoff_borough
0,2019-03-23 20:21:09,2019-03-23 20:27:24,1,1.60,7.0,2.15,0.0,12.95,yellow,credit card,Lenox Hill West,UN/Turtle Bay South,Manhattan,Manhattan
1,2019-03-04 16:11:55,2019-03-04 16:19:00,1,0.79,5.0,0.00,0.0,9.30,yellow,cash,Upper West Side South,Upper West Side South,Manhattan,Manhattan
2,2019-03-27 17:53:01,2019-03-27 18:00:25,1,1.37,7.5,2.36,0.0,14.16,yellow,credit card,Alphabet City,West Village,Manhattan,Manhattan
3,2019-03-10 01:23:59,2019-03-10 01:49:51,1,7.70,27.0,6.15,0.0,36.95,yellow,credit card,Hudson Sq,Yorkville West,Manhattan,Manhattan
4,2019-03-30 13:27:42,2019-03-30 13:37:14,3,2.16,9.0,1.10,0.0,13.40,yellow,credit card,Midtown East,Yorkville West,Manhattan,Manhattan


---
## Actividad 1. Conversión y validación (20 puntos)

### 1.1 Conversión (5 pts)

1. Muestra el tipo de dato de `pickup` y `dropoff`.
2. Convierte ambas columnas a tipo fecha con `pd.to_datetime()`, indicando el **formato explícito** que corresponde a los valores.
3. Verifica que la conversión no generó `NaT`: compara el número de faltantes antes y después.

In [2]:
# Tipo de dato original (al leer el CSV, las fechas vienen como texto: object)
print('Tipos antes de convertir:')
print(taxis[['pickup', 'dropoff']].dtypes)

# Faltantes antes de la conversión (valores nulos en el texto original)
faltantes_antes = taxis[['pickup', 'dropoff']].isna().sum()

# Conversión con formato explícito.
taxis['pickup'] = pd.to_datetime(taxis['pickup'], format='%Y-%m-%d %H:%M:%S')
taxis['dropoff'] = pd.to_datetime(taxis['dropoff'], format='%Y-%m-%d %H:%M:%S')

# Faltantes DESPUÉS de la conversión
faltantes_despues = taxis[['pickup', 'dropoff']].isna().sum()

# Comparación antes vs después
comparacion_nat = pd.DataFrame({
    'faltantes_antes': faltantes_antes,
    'faltantes_despues': faltantes_despues
})
comparacion_nat['nuevos_NaT'] = comparacion_nat['faltantes_despues'] - comparacion_nat['faltantes_antes']
print('\nComparación de faltantes (de', len(taxis), 'filas):')
print(comparacion_nat)

print('\nTipos después de convertir:')
print(taxis[['pickup', 'dropoff']].dtypes)

Tipos antes de convertir:
pickup     object
dropoff    object
dtype: object

Comparación de faltantes (de 6433 filas):
         faltantes_antes  faltantes_despues  nuevos_NaT
pickup                 0                  0           0
dropoff                0                  0           0

Tipos después de convertir:
pickup     datetime64[ns]
dropoff    datetime64[ns]
dtype: object


### 1.2 Periodo de los datos (5 pts)

El dataset debería contener solo viajes de marzo de 2019.

1. Muestra la fecha mínima y máxima de `pickup` y de `dropoff`.
2. Cuenta y muestra los viajes cuyo `pickup` **no** está en marzo de 2019.
3. ¿Hay viajes con `pickup` en marzo y `dropoff` fuera de marzo? Cuéntalos.

In [3]:
print('Fecha minima pickup :', taxis['pickup'].min())
print('Fecha maxima pickup :', taxis['pickup'].max())
print('Fecha minima dropoff:', taxis['dropoff'].min())
print('Fecha maxima dropoff:', taxis['dropoff'].max())

inicio = '2019-03-01'
fin = '2019-04-01'   # límite exclusivo

mask_pickup = (taxis['pickup'] >= inicio) & (taxis['pickup'] < fin)
mask_dropoff = (taxis['dropoff'] >= inicio) & (taxis['dropoff'] < fin)

print('\nPickup dentro del rango :', mask_pickup.sum())
print('Pickup fuera del rango  :', (~mask_pickup).sum())
print('Dropoff dentro del rango:', mask_dropoff.sum())
print('Dropoff fuera del rango :', (~mask_dropoff).sum())

pickup_fuera_marzo = taxis[~mask_pickup]

print('Total fuera de marzo:', len(pickup_fuera_marzo))
print(pickup_fuera_marzo[['pickup', 'dropoff']])

# Viajes con pickup en marzo pero dropoff fuera de marzo
cruzan_mes = taxis[mask_pickup & ~mask_dropoff]
print('Pickup en marzo y dropoff fuera:', len(cruzan_mes))
print(cruzan_mes[['pickup', 'dropoff']])

Fecha minima pickup : 2019-02-28 23:29:03
Fecha maxima pickup : 2019-03-31 23:43:45
Fecha minima dropoff: 2019-02-28 23:32:35
Fecha maxima dropoff: 2019-04-01 00:13:58

Pickup dentro del rango : 6432
Pickup fuera del rango  : 1
Dropoff dentro del rango: 6431
Dropoff fuera del rango : 2
Total fuera de marzo: 1
                  pickup             dropoff
6203 2019-02-28 23:29:03 2019-02-28 23:32:35
Pickup en marzo y dropoff fuera: 1
                 pickup             dropoff
591 2019-03-31 23:43:45 2019-04-01 00:13:58


### 1.3 Cambio de horario (5 pts)

El 10 de marzo de 2019, en Nueva York, las horas entre 02:00 y 02:59 **no existieron**.

1. Filtra los viajes con `pickup` el 10 de marzo y cuenta cuántos hay en cada hora (`.dt.hour` y `groupby()` o `value_counts()`). ¿Hay viajes a las 2:00?
2. Localiza `pickup` en la zona `'America/New_York'` con `tz_localize()` usando `nonexistent='NaT'` y cuenta cuántos valores quedan como `NaT`.

In [4]:
inicio_cambio = '2019-03-10'
fin_cambio = '2019-03-11'

mask_pickup_diezdemarzo = (taxis['pickup'] >= inicio_cambio) & (taxis['pickup'] < fin_cambio)

print('Pickup dentro del 10 de marzo :', mask_pickup_diezdemarzo.sum())
print('Pickup fuera del 10 de marzo  :', (~mask_pickup_diezdemarzo).sum())

pickup_fuera_diezdemarzo = taxis[~mask_pickup_diezdemarzo]
print('Total fuera del 10 de marzo:', len(pickup_fuera_diezdemarzo))

# Viajes del 10 de marzo contados por hora
pickup_10marzo = taxis[mask_pickup_diezdemarzo]

viajes_por_hora = pickup_10marzo['pickup'].dt.hour.value_counts().sort_index()
viajes_por_hora = viajes_por_hora.reindex(range(24), fill_value=0)

print('\nViajes por hora (10 de marzo de 2019):')
print(viajes_por_hora)

# ¿Hay viajes a las 2:00?
viajes_2am = (pickup_10marzo['pickup'].dt.hour == 2).sum()
print('\nViajes con pickup entre 02:00 y 02:59:', viajes_2am)

pickup_ny = taxis['pickup'].dt.tz_localize('America/New_York', nonexistent='NaT')

print(pickup_ny.dtype)   # datetime64[ns, America/New_York]
print('NaT después de localizar:', pickup_ny.isna().sum(), 'de', len(pickup_ny))

Pickup dentro del 10 de marzo : 185
Pickup fuera del 10 de marzo  : 6248
Total fuera del 10 de marzo: 6248

Viajes por hora (10 de marzo de 2019):
pickup
0     11
1      7
2      0
3      6
4      7
5      1
6      2
7      5
8      7
9      9
10    13
11    11
12    17
13     8
14    13
15     9
16    10
17    14
18     7
19     3
20     4
21     3
22    10
23     8
Name: count, dtype: int64

Viajes con pickup entre 02:00 y 02:59: 0
datetime64[ns, America/New_York]
NaT después de localizar: 0 de 6433


### 1.4 Conclusión (5 pts)

¿Qué problemas temporales encontraste en esta actividad? ¿Qué indica el resultado de 1.3 sobre la zona horaria en que están registrados los datos? Responde con cifras.

**Respuesta 1.4:**

De entrada al realizar la conversión no encontré ningún NaT lo cual me indica que el formato era consistente desde el principio. Otro problema es que supuestamente el dataset es unicamente de marzo de 2019, sin embargo se encontró un viaje que inicia el 28 de febrero y otro que inicia el 31 de marzo pero termina el 1 de abril lo cual podría afectar y finalmente, el cambio de horario se refleja en los datos, ya que no existen registros de viajes entre las 02:00 y 02:59 del 10 de marzo

---
## Actividad 2. Duraciones y anomalías (25 puntos)

### 2.1 Duración de cada viaje (6 pts)

1. Crea la columna `duracion` = `dropoff` − `pickup`.
2. Crea `duracion_min` con la duración en **minutos** (usa `.dt.total_seconds()`).
3. Muestra el resumen estadístico de `duracion_min` con `describe()`.

In [5]:
# 1. Duración como Timedelta
taxis['duracion'] = taxis['dropoff'] - taxis['pickup']

# 2. Duración en minutos
taxis['duracion_min'] = taxis['duracion'].dt.total_seconds() / 60

# 3. Resumen estadístico
print(taxis['duracion_min'].describe())

count    6433.000000
mean       14.349617
std        11.643892
min         0.000000
25%         6.500000
50%        10.900000
75%        18.516667
max       107.666667
Name: duracion_min, dtype: float64


### 2.2 Duraciones anómalas (6 pts)

Cuenta los viajes con:

- Duración **negativa** (termina antes de iniciar).
- Duración **cero**.
- Duración **menor a 1 minuto** (incluye los de duración cero).
- Duración **mayor a 3 horas**.

Muestra los viajes de duración cero con las columnas `pickup`, `dropoff`, `distance`, `fare` y `total`.

In [6]:
neg = (taxis['duracion_min'] < 0).sum()
cero = (taxis['duracion_min'] == 0).sum()
menor_1min = (taxis['duracion_min'] < 1).sum()
mayor_3h = (taxis['duracion_min'] > 180).sum()

print('Duración negativa      :', neg)
print('Duración cero          :', cero)
print('Duración menor a 1 min :', menor_1min)
print('Duración mayor a 3 h   :', mayor_3h)

# Viajes de duración cero
viajes_cero = taxis[taxis['duracion_min'] == 0]
print('\nViajes de duración cero:')
print(viajes_cero[['pickup', 'dropoff', 'distance', 'fare', 'total']])

Duración negativa      : 0
Duración cero          : 6
Duración menor a 1 min : 72
Duración mayor a 3 h   : 0

Viajes de duración cero:
                  pickup             dropoff  distance  fare  total
1690 2019-03-22 06:24:14 2019-03-22 06:24:14       0.0  72.0   72.0
5493 2019-03-06 12:14:00 2019-03-06 12:14:00       0.0  10.0   10.0
5624 2019-03-11 14:04:50 2019-03-11 14:04:50       0.0   2.5    3.3
5638 2019-03-13 12:22:00 2019-03-13 12:22:00       0.0  10.0   10.0
6083 2019-03-19 15:34:00 2019-03-19 15:34:00       0.0   5.0    5.0
6356 2019-03-01 16:58:23 2019-03-01 16:58:23       0.0   3.0    4.8


### 2.3 Velocidad promedio (6 pts)

La velocidad promedio de un viaje, en millas por hora, es `distance / (duracion_min / 60)`.

1. Crea la columna `velocidad_mph` **solo para viajes con duración mayor que cero**; en los demás debe quedar como faltante (`NaN`). Sugerencia: calcula la velocidad para todos los viajes y conserva solo los valores válidos con `.where(condición)`, que deja `NaN` donde la condición es `False`.
2. Cuenta los viajes con velocidad mayor a 60 mph (velocidad poco plausible en tráfico urbano).
3. Muestra los 5 viajes con mayor velocidad, ordenados de mayor a menor, con `pickup`, `dropoff`, `duracion_min`, `distance`, `velocidad_mph` y `fare`.

In [7]:
# Velocidad en mph (solo para duración > 0
velocidad = taxis['distance'] / (taxis['duracion_min'] / 60)
taxis['velocidad_mph'] = velocidad.where(taxis['duracion_min'] > 0)

# Viajes con velocidad mayor a 60 mph
mas_60 = (taxis['velocidad_mph'] > 60).sum()
print('Viajes con velocidad > 60 mph:', mas_60)

# los 5 viajes más rápidos
top5 = taxis.sort_values('velocidad_mph', ascending=False).head(5)
print(top5[['pickup', 'dropoff', 'duracion_min', 'distance', 'velocidad_mph', 'fare']])

Viajes con velocidad > 60 mph: 11
                  pickup             dropoff  duracion_min  distance  \
770  2019-03-02 03:16:59 2019-03-02 03:17:06      0.116667      9.40   
6089 2019-03-17 23:16:34 2019-03-17 23:16:52      0.300000     10.10   
491  2019-03-07 07:11:33 2019-03-07 07:11:39      0.100000      1.60   
5284 2019-03-22 06:24:02 2019-03-22 06:24:11      0.150000      1.70   
3241 2019-03-24 17:19:27 2019-03-24 17:19:29      0.033333      0.34   

      velocidad_mph  fare  
770     4834.285714   2.5  
6089    2020.000000  12.0  
491      960.000000   2.5  
5284     680.000000   2.5  
3241     612.000000  65.0  


### 2.4 Criterio de exclusión (7 pts)

Con base en 1.2, 2.2 y 2.3, define qué viajes excluirías del análisis y por qué. Considera:

a) ¿Todos los viajes de menos de 1 minuto son errores? Observa su distancia y tarifa.

b) ¿Qué umbrales (duración, velocidad, periodo) usarás? Justifícalos.

**Respuesta 2.4:**

a) ¿Todos los viajes de menos de 1 minuto son errores? Observa su distancia y tarifa.

No necesariamente todos los viajes menores a un minuto son errores ya que algunos podrían corresponder a trayectos muy cortos. Sin embargo, debido a la presencia de registros con distancia cero y otros con combinaciones de distancia y duración que generan velocidades físicamente poco plausibles, creo que si debemos usarlo como criterio de exclusion

b) ¿Qué umbrales (duración, velocidad, periodo) usarás? Justifícalos.

**Periodo:** solo viajes con pickup entre el 1 y el 31 de marzo de 2019. En 1.2 encontré 1 viaje fuera de ese rango (28 de febrero), que no pertenece al periodo de estudio.
**Duración mínima de 1 minuto:** excluye los 72 viajes de menos de 1 minuto, por lo que se ve arriba: distancias de 0 millas y velocidades imposibles. Es una decisión de criterio
**Duración máxima de 3 horas:** en 2.2 hay 0 viajes con duración mayor a 3 horas, así que no excluye ninguno, sin embargo lo mantendría como un posible outlier, dado que sería poco probable que un viaje por la ciudad dure más de ese tiempo
**Velocidad máxima de 60 mph:** en 2.3 hay 11 viajes por encima de este umbral. Es un límite para tráfico urbano por lo que los casos que superen este umbral podrían ser outliers que distorsionen mis datos y conclusiones

---
## Actividad 3. Componentes y patrones (25 puntos)

### 3.1 Extraer componentes (6 pts)

Crea las columnas:

- `hora`: hora de `pickup`.
- `dia_num`: día de la semana de `pickup` (lunes = 0).
- `dia_nombre`: nombre del día de `pickup`.
- `fin_semana`: `True` si es sábado o domingo.
- `fecha`: solo la fecha (sin hora) de `pickup`.
- `cruza_medianoche`: `True` si la **fecha** de `dropoff` es distinta de la de `pickup`.

Cuenta cuántos viajes cruzan la medianoche.

In [8]:
taxis['hora'] = taxis['pickup'].dt.hour
taxis['dia_num'] = taxis['pickup'].dt.dayofweek          # lunes = 0, domingo = 6
taxis['dia_nombre'] = taxis['pickup'].dt.day_name()
taxis['fin_semana'] = taxis['dia_num'] >= 5              # sábado (5) o domingo (6)
taxis['fecha'] = taxis['pickup'].dt.date
taxis['cruza_medianoche'] = taxis['dropoff'].dt.date != taxis['pickup'].dt.date

print('Viajes que cruzan la medianoche:', taxis['cruza_medianoche'].sum())

Viajes que cruzan la medianoche: 50


### 3.2 Patrón por hora del día (7 pts)

Con `groupby('hora')`, calcula para cada hora: el **número de viajes** y la **mediana** de `duracion_min`. Indica la hora con más viajes y la hora con menos.

Usa la mediana (no la media) porque la duración tiene valores extremos.

In [9]:
por_hora = taxis.groupby('hora').agg(
    viajes=('hora', 'size'),
    mediana_duracion_min=('duracion_min', 'median')
)
print(por_hora)

hora_max = por_hora['viajes'].idxmax()
hora_min = por_hora['viajes'].idxmin()

print(f"\nHora con más viajes  : {hora_max}:00 ({por_hora.loc[hora_max, 'viajes']} viajes)")
print(f"Hora con menos viajes: {hora_min}:00 ({por_hora.loc[hora_min, 'viajes']} viajes)")

      viajes  mediana_duracion_min
hora                              
0        205             10.733333
1        110              9.633333
2        101              9.183333
3         67              9.583333
4         57             10.500000
5         51              8.700000
6        142              7.833333
7        221              9.583333
8        313             11.316667
9        319             11.616667
10       325             11.883333
11       295             12.033333
12       334             12.775000
13       316             10.558333
14       360             11.841667
15       330             12.058333
16       336             11.583333
17       388             11.975000
18       417             11.100000
19       406             10.650000
20       368              9.258333
21       355             11.083333
22       321              9.816667
23       296             10.575000

Hora con más viajes  : 18:00 (417 viajes)
Hora con menos viajes: 5:00 (51 viajes)


### 3.3 Patrón por día de la semana (6 pts)

Calcula por día de la semana el **número de viajes** y la **mediana** de `duracion_min` y de `fare`. Presenta los días en orden de lunes a domingo (sugerencia: agrupa por `['dia_num', 'dia_nombre']`).

Toma en cuenta que marzo de 2019 no tiene el mismo número de cada día de la semana: calcula también el **promedio de viajes por fecha**, dividiendo entre el número de fechas distintas de ese día de la semana (usa la columna `fecha` y `.nunique()`).

In [10]:
por_dia = taxis.groupby(['dia_num', 'dia_nombre']).agg(
    viajes=('dia_num', 'size'),
    mediana_duracion_min=('duracion_min', 'median'),
    mediana_fare=('fare', 'median'),
    fechas_distintas=('fecha', 'nunique')
)

por_dia['viajes_por_fecha'] = (por_dia['viajes'] / por_dia['fechas_distintas']).round(1)

print(por_dia)

                    viajes  mediana_duracion_min  mediana_fare  \
dia_num dia_nombre                                               
0       Monday         708             10.541667           9.5   
1       Tuesday        825             11.216667           9.5   
2       Wednesday      966             11.291667           9.5   
3       Thursday       905             11.766667           9.5   
4       Friday        1115             11.150000           9.5   
5       Saturday      1046             10.466667           9.0   
6       Sunday         868              9.983333           9.0   

                    fechas_distintas  viajes_por_fecha  
dia_num dia_nombre                                      
0       Monday                     4             177.0  
1       Tuesday                    4             206.2  
2       Wednesday                  4             241.5  
3       Thursday                   5             181.0  
4       Friday                     5             223.0  
5     

### 3.4 Interpretación (6 pts)

a) ¿Qué patrones observas por hora y por día? ¿Coinciden la hora con más viajes y la hora con mayor duración mediana? ¿Qué podría explicarlo?

b) ¿Por qué fue necesario calcular el promedio de viajes por fecha en 3.3 en lugar de comparar solo el total por día de la semana?

**Respuesta 3.4:**

a) Por hora observo que el número de viajes es muy bajo de madrugada: la hora con menos viajes es la 5:00, con 51, y entre las 3:00 y las 5:00 hay entre 51 y 67 viajes por hora. A partir de las 6:00 (142) sube rápido, hasta 313 a las 8:00, y se mantiene entre unos 295 y 360 durante el día. Alcanza su máximo por la tarde: la hora con más viajes es la 18:00, con 417, y las horas de 17:00 a 19:00 son las más altas (388, 417 y 406). Después baja gradualmente hasta 296 a las 23:00.

Por otro lado, por día se observa qe el viernes concentra más viajes en total (1,115), seguido del sábado (1,046), y el lunes tiene menos (708). Los viajes de fin de semana son un poco más cortos y baratos: la mediana de duración es de 10.47 min el sábado y 9.98 min el domingo, frente a 10.54 - 11.77 min de lunes a viernes, y la mediana de tarifa es de 9.00 el fin de semana contra 9.50 entre semana.

**¿Coinciden la hora con más viajes y la de mayor duración mediana?** No. La hora con más viajes es la 18:00 (mediana de 11.10 min), mientras que la mayor duración mediana ocurre a las 12:00 (12.78 min, con 334 viajes), y la menor a las 6:00 (7.83 min, con 142 viajes). Que haya más viajes no implica que sean más largos. Una explicación posible es que a mediodía hay más tráfico y actividad en la ciudad y por eso los viajes tardan más, mientras que a las 6:00 las calles están más libres y los viajes son cortos. La hora de mayor demanda (18:00) coincide con la salida del trabajo, pero su duración mediana no es la mayor. Mis hipotesis son que los datos no incluyen tráfico, pero no tenemos esos datos. También las medianas varían poco (entre 7.83 y 12.78 min), por lo que no hay tampoco tanta diferencia.

b) Porque marzo de 2019 no tiene el mismo número de cada día de la semana: lunes, martes y miércoles tienen 4 fechas, mientras que jueves, viernes, sábado y domingo tienen 5. Si comparo solo los totales, los días con más fechas parecen tener más viajes solo porque hay más días de ese tipo.

---
## Actividad 4. Dataset limpio y comparación (20 puntos)

### 4.1 Dataset limpio (6 pts)

Crea `taxis_limpio` aplicando el criterio que definiste en 2.4. Imprime cuántos viajes se excluyeron por cada condición y en total, y verifica que `taxis_limpio` cumple todas las condiciones.

In [11]:
# Condiciones de exclusión de acuerdo a 2.4
c_periodo = ~((taxis['pickup'] >= '2019-03-01') & (taxis['pickup'] < '2019-04-01'))  # fuera de marzo 2019
c_corta   = taxis['duracion_min'] < 1      # duración < 1 min
c_larga   = taxis['duracion_min'] > 180    # duración > 3 h
c_veloz   = taxis['velocidad_mph'] > 60    # velocidad poco probable

print('Fuera de marzo 2019 :', c_periodo.sum())
print('Duración < 1 min    :', c_corta.sum())
print('Duración > 3 h      :', c_larga.sum())
print('Velocidad > 60 mph  :', c_veloz.sum())

# Un viaje puede cumplir varias condiciones
excluir = c_periodo | c_corta | c_larga | c_veloz
print('\nTotal excluidos (sin doble conteo):', excluir.sum())

# Dataset limpio: se conservan los que no cumplen ninguna condición
taxis_limpio = taxis[~excluir].copy()
print('Antes:', len(taxis), '| Después:', len(taxis_limpio))

# Verificación
assert taxis_limpio['pickup'].between('2019-03-01', '2019-03-31 23:59:59').all()
assert (taxis_limpio['duracion_min'] >= 1).all()
assert (taxis_limpio['duracion_min'] <= 180).all()
assert (taxis_limpio['velocidad_mph'] <= 60).all()
print('Verificación superada: taxis_limpio cumple todas las condiciones.')
print("Duración mínima:", taxis_limpio["duracion_min"].min(), "minutos")
print("Duración máxima:", taxis_limpio["duracion_min"].max(), "minutos")
print("Velocidad máxima:", taxis_limpio["velocidad_mph"].max(), "mph")

Fuera de marzo 2019 : 1
Duración < 1 min    : 72
Duración > 3 h      : 0
Velocidad > 60 mph  : 11

Total excluidos (sin doble conteo): 74
Antes: 6433 | Después: 6359
Verificación superada: taxis_limpio cumple todas las condiciones.
Duración mínima: 1.0 minutos
Duración máxima: 107.66666666666667 minutos
Velocidad máxima: 54.174757281553404 mph


### 4.2 Comparación entre periodos (6 pts)

Con `taxis_limpio`, compara los viajes que inician entre las **7:00 y las 9:59** en **días hábiles** contra los del **mismo horario en fin de semana**. Para cada grupo reporta: número de viajes, mediana de `duracion_min` y mediana de `velocidad_mph`.

In [12]:
# Ventana 7:00–9:59 con la hora de pickup
en_ventana = taxis_limpio['pickup'].dt.hour.between(7, 9)
ventana = taxis_limpio[en_ventana].copy()

# Día hábil vs fin de semana
ventana['grupo'] = ventana['pickup'].dt.dayofweek.map(
    lambda d: 'Fin de semana' if d >= 5 else 'Día hábil'
)

comparacion = ventana.groupby('grupo').agg(
    viajes=('grupo', 'size'),
    mediana_duracion_min=('duracion_min', 'median'),
    mediana_velocidad_mph=('velocidad_mph', 'median')
).round(2)
print(comparacion)

               viajes  mediana_duracion_min  mediana_velocidad_mph
grupo                                                             
Día hábil         683                 12.03                   9.10
Fin de semana     161                  7.77                  13.47


### 4.3 Registro de decisión (8 pts)

Redacta el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Conversión**: formato utilizado y verificación.
- **Zona horaria**: en qué zona están los datos y cómo lo verificaste.
- **Criterio de exclusión**: condiciones, umbrales y justificación.
- **Resultado**: viajes antes, después y excluidos.
- **Hallazgo de 4.2** y **limitaciones** del análisis.

**Reporte de registro de limpieza: viajes de taxi de Nueva York, marzo de 2019**

**Conversión.** Las columnas pickup y dropoff venían como texto y se convirtieron a datetime64 con pd.to_datetime() con formato explícito . Se verificó comparando los faltantes antes y después: 0 antes y 0 después en ambas columnas, es decir, 0 NaT nuevos de 6,433 filas. La conversión fue completa y no se perdió ningún dato.

**Zona horaria.** Los datos están en hora local de Nueva York. Lo verifiqué con el cambio de horario del 10 de marzo de 2019: ese día hay 185 viajes y ninguno con pickup entre las 02:00 y las 02:59, la hora que no existió. Además, al localizar pickup con tz_localize('America/New_York, nonexistent=NaT)` quedaron 0 NaT de 6,433 valores. La ausencia de registros entre las 02:00 y 02:59 es consistente con que los timestamps estén en hora local de Nueva York

**Criterio de exclusión.** Se excluyeron los viajes que cumplían al menos una de estas condiciones:

- pickup fuera del 1 al 31 de marzo de 2019: 1 viaje (28 de febrero).
- Duración menor a 1 minuto: 72 viajes (6 con duración cero, ninguno con duración negativa). Su duración no es confiable: los 6 de duración cero tienen 0.0 millas pero tarifas distintas (2.50 a 72), y los 5 viajes más rápidos de todo el dataset duran menos de 1 minuto (por ejemplo, 9.4 millas en 0.12 minutos, o 4,834 mph).
- Duración mayor a 3 horas: 0 viajes (la máxima observada es de 107.7 minutos); se mantiene como salvaguarda.
- Velocidad mayor a 60 mph: 11 viajes, límite físico razonable en tráfico urbano.

El umbral de 1 minuto es una decisión de criterio por considerarse un outlier

**Resultado.** Antes: 6,433 viajes. Excluidos: 74 (1.15 %), sin doble conteo, porque 10 de los 11 viajes de más de 60 mph ya estaban entre los de menos de 1 minuto. Después: 6,359 viajes. Se verificó que taxis_limpio cumple todas las condiciones: la duración mínima es 1.0 minuto y la velocidad máxima es 54.2 mph.

**Hallazgo de 4.2.** Entre las 7:00 y las 9:59, los días hábiles tienen 683 viajes, con mediana de 12.03 min y 9.10 mph. El fin de semana tiene 161 viajes, con mediana de 7.77 min y 13.47 mph. Los viajes de día hábil duran alrededor de 55 % más y avanzan a menor velocidad, lo que es coherente con la congestión de la hora pico entre semana. Como la muestra incluye 21 días hábiles y 10 de fin de semana, el promedio por día es de unos 32.5 viajes contra 16.1.

**Limitaciones.**
- Es una muestra de un solo mes (6,433 viajes), y el grupo de fin de semana tiene solo 161 viajes en la ventana, así que sus medianas son menos estables.
- La comparación no controla la distancia ni la zona de origen o destino; parte de la diferencia de duración podría deberse a que se hacen trayectos distintos, no solo a la congestión.
- Los umbrales (1 minuto, 3 horas, 60 mph) son criterio propio y podría ser un tanto subjetivo con otros umbrales cambiaría el número de excluidos. Creo que también depende del país y la movilidad ya que me baso en mi experiencia pero quizá en Nueva York la movilidad sea diferente.
- La zona horaria se verificó con un solo indicio (la hora 2:00 del 10 de marzo); no se comprobó de otra forma.



---
## Reflexión final (10 puntos)

1. ¿Qué ventajas tuvo indicar el formato explícito en 1.1 frente a dejar que pandas lo infiera? ¿En qué situación la inferencia habría producido errores silenciosos?
2. Supón que `pickup` se hubiera registrado en UTC y `dropoff` en hora local de Nueva York. ¿Qué efecto tendría en las duraciones calculadas? ¿Cómo lo habrías detectado con las herramientas de esta actividad?
3. Si se entrenara un modelo para predecir la duración de un viaje, ¿qué problemas causarían las anomalías temporales que encontraste si no se limpiaran?

**Respuesta final:**

1.  Indicar el formato hace que pandas interprete cada valor exactamente como yo lo declaro, sin adivinar, y si algún valor no coincide lanza un error en lugar de pasar inadvertido. Con ese formato, las 6,433 filas de pickup y de dropoff se convirtieron sin NaT nuevos (0 antes, 0 después), y eso lo pude afirmar con certeza. También evita que el resultado dependa de la versión de pandas o de la configuración del equipo, así que el notebook es reproducible.

La inferencia habría producido errores silenciosos en fechas ambiguas en donde pandas hubiese "asumido" la fecha y quizá no me hubiera dado cuenta.

2.  En marzo de 2019, Nueva York está 5 horas detrás de UTC hasta el 10 de marzo y 4 horas detrás desde el cambio de horario. Si pickup estuviera en UTC y dropoff en hora local, la resta dropoff  pickup estaría corrida entre 4 y 5 horas: casi todas las duraciones saldrían negativa y además con dos grupos distintos según si el viaje ocurrió antes o después del 10 de marzo.

Lo habría detectado con las herramientas de esta actividad:
- En 2.1, describe() mostraría una duración mínima, mediana y promedio negativos, cuando en mis datos la mediana es de 10.9 min y el mínimo es 0.
- En 2.2, el conteo de duraciones negativas sería cercano a 6,433, en lugar de los 0 que obtuve.
- En 1.2, el rango de pickup estaría desplazado respecto al de dropoff (varias horas más tarde), cuando en mis datos ambos cubren prácticamente el mismo periodo.
- En 1.3, la prueba del 10 de marzo mostraría viajes a las 2:00 en la columna que estuviera en UTC, ya que UTC no tiene ese salto.

3.  Si entrenara un modelo para predecir la duración sin limpiar los datos, las anomalías le enseñarían relaciones falsas. Hay 72 viajes de menos de 1 minuto (6 con duración cero) y 11 con más de 60 mph, y en los casos extremos la distancia y la duración son incompatibles.Como la distancia es uno de los principales factores de la duración, estos registros deformarían esa relación y el modelo aprendería que viajes largos pueden durar segundos.


---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesión.